frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -45,186 +46,246 @@
+    for (String nodeName : filteredNodes) {
+      candidateNodes.add(allNodes.get(nodeName));
+    }
+    Collections.sort(candidateNodes);
+    for (int i = 0; i < requestNumber && i < candidateNodes.size(); i++) {
+      Node node = candidateNodes.get(i);
+      Long gpuAttribute = requestResource.getGpuAttribute();
+      if (gpuAttribute == 0) {
+        gpuAttribute = selectCandidateGpuAttribute(node, requestResource.getGpuNumber());
+      }
+      result.addSelection(node.getHost(), gpuAttribute, node.getAvailableResource().getPortRanges());
     }
+    return result;
+  }
 
-    if (selectionResult != null) {
-      LOGGER.logInfo(
-          "select: Found a SelectionResult satisfies the Request: SelectionResult: [%s]",
-          selectionResult);
-    } else {
-      LOGGER.logWarning(
-          "select: Cannot found a SelectionResult satisfies the Request: " +
-              "Check whether the Request can be relaxed to RM");
-      String notRelaxLogPrefix = "select: The Request cannot be relaxed to RM: Reason: ";
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        throw new NotAvailableException(
-            String.format(notRelaxLogPrefix +
-                    "NodeGpuType is specified: Request NodeGpuType: [%s]",
-                requestNodeGpuType));
+  public synchronized SelectionResult selectSingleNode(ResourceDescriptor requestResource,
+      String taskRoleName) throws NotAvailableException {
+    SelectionResult results = select(requestResource, taskRoleName);
+    // Random pick a host from the result set to avoid conflicted requests from concurrent container requests
+    // from different jobs
+    ResourceDescriptor optimizedRequestResource = results.getOptimizedResource();
+    String candidateNode = results.getNodeHosts().get(CommonUtils.getRandomNumber(0, results.getNodeHosts().size() - 1));
+    optimizedRequestResource.setGpuAttribute(results.getGpuAttribute(candidateNode));
+
+    // re-create single node result object.
+    SelectionResult result = new SelectionResult();
+    result.addSelection(candidateNode, results.getGpuAttribute(candidateNode), results.getOverlapPorts());
+    result.setOptimizedResource(optimizedRequestResource);
+
+    return result;
+  }
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String taskRoleName)
+      throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: TaskRole: [%s] Resource: [%s]", taskRoleName, requestResource);
+    String requestNodeLabel = requestManager.getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
+    String requestNodeGpuType = requestManager.getTaskPlatParams().get(taskRoleName).getTaskNodeGpuType();
+    int pendingTaskNumber = statusManager.getUnassociatedTaskCount(taskRoleName);
+    List<ValueRange> reUsePorts = null;
+
+    // Prefer to use previous successfully allocated ports. if no successfully ports, try to re-use the "Requesting" ports.
```

Comment:
>allocated [](start = 43, length = 9)

rename allocate to select in this file

Reply:
this line not need change, other line already changed. 

---
In reply to: [175706219](https://github.com/Microsoft/pai/pull/318#discussion_r175706219) [](ancestors = 175706219)

Reply:
But you also used 
reUsePorts = previousRequestedPorts.get(taskRoleName);

---
In reply to: [175728409](https://github.com/Microsoft/pai/pull/318#discussion_r175728409) [](ancestors = 175728409,175706219)